L73-java-binderchannel/wireformat.md

```diff
@@ -25,38 +25,60 @@ There are two types of transactions, control and stream.
 For general transport lifecycle, using one of 1000 reserved transaction codes.
 
 *   Setup Transport
-    *   Transaction code: BinderTransport.SETUP_TRANSPORT
+    *   Transaction code: BinderTransport.SETUP_TRANSPORT (1)
     *   From ClientTransport to host binder, and from ServerTransport to client
         binder.
 *   Shutdown Transport
-    *   Transaction code: BinderTransport.SHUTDOWN_TRANSPORT
+    *   Transaction code: BinderTransport.SHUTDOWN_TRANSPORT (2)
     *   From ClientTransport to server binder, or ServerTransport to client
         binder.
 *   Acknowledge bytes.
-    *   Transaction code: BinderTransport.ACKNOWLEDGE_BYTES
+    *   Transaction code: BinderTransport.ACKNOWLEDGE_BYTES (3)
     *   From either transport to its peer's binder. Reports reception of
         transaction data for flow control.
 *   Ping
-    *   Transaction code: BinderTransport.PING
-    *   From ClientTransport to server binder.
+    *   Transaction code: BinderTransport.PING (4)
+    *   Currently only sent from ClientTransport to server binder.
 *   Ping Response:
-    *   Transaction code: BinderTransport.PING_RESPONSE
-    *   From SeverTransport to client binder, in response to a ping.
+    *   Transaction code: BinderTransport.PING_RESPONSE (5)
+    *   From SeverTransport to client binder or ClientTransport to server binder,
+        in response to a ping message.
+
+If either side of a transport receives an unrecognized control transacton,
+it will shutdown the transport gracefully.
 
 #### Transaction Data
 
 ```
-version = int; (* Currently always 1 *)
+major version = int;
+minor version = int;
+protocol extension flags = int;
+version extras = minor version, protocol extension flags
```

Comment:
Thinking about the server auto-downgrading to an older major version, made me think about if we missed anything else. We don't have any magic bytes right now. Do we need them? I'm thinking they don't add value for us, but just wanted to check.

We could add them to `version extras`, but it sort of seems like magic bytes are unlikely to be helpful if `version extras` is after `binder`.

Reply:
Oh good catch, we don't need something new, since we effectively have an equivalent in the intent we use to bind in the first place. "grpc.io.action.BIND". Adding that.